utilities/transactions/transaction_base.cc

```diff
@@ -632,6 +632,17 @@ void TransactionBaseImpl::TrackKey(TransactionKeyMap* key_map, uint32_t cfh_id,
                                    const std::string& key, SequenceNumber seq,
                                    bool read_only, bool exclusive) {
   auto& cf_key_map = (*key_map)[cfh_id];
+#ifdef __cpp_lib_unordered_map_try_emplace
+  // use c++17's try_emplace if available, to avoid rehashing the key
+  // in case it is not already in the map
+  auto result = cf_key_map.try_emplace(key, TransactionKeyMapInfo(seq));
```

Comment:
Actually, can't we just use regular `emplace` in this case?

From https://en.cppreference.com/w/cpp/container/map/try_emplace:

> Unlike insert or emplace, these functions do not move from rvalue arguments if the insertion does not happen...

In this case though, key is not an rvalue reference, and the value is just a temporary, so we don't care if it got moved or not, so we're okay if that's left in an unspecified state.

Reply:
IMHO emplace will always perform a dynamic memory allocation, regardless of whether the key is already in the map or not.
try_emplace however only allocates memory if the key isnt yet contained in the map.
So if you expect lots of duplicates, emplace may actually perform worse than a combination of find and insert/emplace.
try_emplace seems to be a good general solution, in case there are many duplicates and also in case there arent.

Reply:
Makes sense.

I think you can avoid constructing `TransactionKeyMapInfo` in the failure case if you just pass in seq.